nexus/db-queries/src/db/pool.rs

```diff
@@ -3,108 +3,143 @@
+use tokio::sync::watch;
 
 pub use super::pool_connection::DbConnection;
 
+type QorbConnection = async_bb8_diesel::Connection<DbConnection>;
+type QorbPool = qorb::pool::Pool<QorbConnection>;
+
 /// Wrapper around a database connection pool.
 ///
 /// Expected to be used as the primary interface to the database.
 pub struct Pool {
-    pool: bb8::Pool<ConnectionManager<DbConnection>>,
+    inner: QorbPool,
 }
 
-impl Pool {
-    pub fn new(log: &slog::Logger, db_config: &DbConfig) -> Self {
-        // Make sure diesel-dtrace's USDT probes are enabled.
-        usdt::register_probes().expect("Failed to register USDT DTrace probes");
-        Self::new_builder(log, db_config, bb8::Builder::new())
-    }
+// Provides an alternative to the DNS resolver for cases where we want to
+// contact the pool directly.
+struct SingleHostResolver {
+    tx: watch::Sender<AllBackends>,
+}
 
-    pub fn new_failfast_for_tests(
-        log: &slog::Logger,
-        db_config: &DbConfig,
-    ) -> Self {
-        Self::new_builder(
-            log,
-            db_config,
-            bb8::Builder::new()
-                .connection_timeout(std::time::Duration::from_millis(1)),
-        )
+impl SingleHostResolver {
+    fn new(config: &DbConfig) -> Self {
+        let backends = Arc::new(BTreeMap::from([(
+            backend::Name::new("singleton"),
+            backend::Backend { address: config.url.address() },
+        )]));
+        let (tx, _rx) = watch::channel(backends.clone());
+        Self { tx }
     }
+}
 
-    fn new_builder(
-        log: &slog::Logger,
-        db_config: &DbConfig,
-        builder: bb8::Builder<ConnectionManager<DbConnection>>,
-    ) -> Self {
-        let url = db_config.url.url();
-        let log = log.new(o!(
-            "database_url" => url.clone(),
-            "component" => "db::Pool"
-        ));
-        info!(&log, "database connection pool");
-        let error_sink = LoggingErrorSink::new(log);
-        let manager = ConnectionManager::<DbConnection>::new(url);
-        let pool = builder
-            .connection_customizer(Box::new(
-                super::pool_connection::ConnectionCustomizer::new(),
-            ))
-            .error_sink(Box::new(error_sink))
-            .build_unchecked(manager);
-        Pool { pool }
+impl Resolver for SingleHostResolver {
+    fn monitor(&mut self) -> watch::Receiver<AllBackends> {
+        self.tx.subscribe()
     }
+}
 
-    /// Returns a reference to the underlying pool.
-    pub fn pool(&self) -> &bb8::Pool<ConnectionManager<DbConnection>> {
-        &self.pool
-    }
+fn make_dns_resolver(
+    bootstrap_dns: Vec<SocketAddr>,
+) -> qorb::resolver::BoxedResolver {
+    Box::new(DnsResolver::new(
+        service::Name(internal_dns::ServiceName::Cockroach.srv_name()),
+        bootstrap_dns,
+        DnsResolverConfig {
+            query_interval: tokio::time::Duration::from_secs(10),
+            hardcoded_ttl: Some(std::time::Duration::from_secs(60)),
```

Comment:
This `hardcoded_ttl` config is load-bearing. The default is `None`, and the `DnsResolver` uses a TTL from the DNS records themselves to identify the the records as present or not. **However**, all our DNS records are being saved with a TTL of "zero", so they are immediately discarded.